Dockerfile

```diff
@@ -1,34 +1,67 @@
-# use multi-stage builds to independently pull dependency versions
-ARG quilc_version=1.20.0
-ARG qvm_version=1.17.1
-ARG python_version=3.7
+# syntax=docker/dockerfile:1.3
+ARG PYTHON_VERSION=3.8
```

Comment:
Any reason we wouldn't use something more recent?

Reply:
Nope. What version do you prefer?

Reply:
If it makes no difference, I would prefer `3.11` since it is the newest and gives us the most lead time before it gets deprecated and we have to update this again.

Reply:
The version of numpy in the lockfile is 1.21.1 and the scipy version is 1.61. I am having difficulties installing these on python base images with a version >=3.10.

For python 3.10, we hit a problem with scipy 1.6.1

```
#0 260.7 Note: This error originates from the build backend, and is likely not a problem with poetry but with scipy (1.6.1) not supporting PEP 517 builds. You can verify this by running 'pip wheel --use-pep517 "scipy (==1.6.1)"'.
```

For python 3.11, we hit a problem with numpy 1.21.1

```
#0 108.0 Note: This error originates from the build backend, and is likely not a problem with poetry but with numpy (1.21.1) not supporting PEP 517 builds. You can verify this by running 'pip wheel --use-pep517 "numpy (==1.21.1)"'.
```

Glancing at the tests, it looks like we test 3.7, 3.8, and 3.9 so this would not be caught by the tests.

Furthermore, I tried increasing the base numpy version to 1.22, however that conflicts with python 3.7 support.

As such, I have put the Dockerfile python version to 3.9, which is the highest version compatible with the poetry.lock file.

This probably deserves a follow-on issue.

Reply:
Ah, strange. Thanks for creating #1606. It makes sense to address that separately from this. 3.9 should be good for a while.

Reply:
> … it looks like we test 3.7, 3.8, and 3.9
Numpy's [version deprecation policy](https://numpy.org/neps/nep-0029-deprecation_policy.html) calls for dropping support for 3.7 support in December 2021 and 3.8 in April of this year. Given how integrated `pyquil` is with the rest of the "pydata stack" (it's listed on numpy.org !) we should probably follow their schedule.

Reply:
They also dropped numpy 1.21 support this month.